### PR TITLE
Drone CI, static builds, Windows builds

### PR DESCRIPTION
**Pull request by jagerman**
- Adds CI for the project.
- Adds static dep builds
- Makes windows build work in mingw (though doesn't appear to *work* yet).
- Fixes various problems caught by the CI jobs.

This adds a suite of Drone CI builds, and fixes various issues where things weren't building.

This includes full static builds, via the same -DBUILD_STATIC_DEPS=ON mode we have in other software.

Finally this adds a Windows build, including lots of ugly unholiness brought in from hell, including a suitably named in constexpr constant to help tell when we are in such an environment.

### CODE REVIEW

File: include/quic/gnutls_crypto.hpp [hunk 1]
```diff
@@ -33,46 +33,46 @@ namespace oxen::quic
...
-        // returns truew if path is not empty OR mem has a value set
-        inline explicit operator bool() const noexcept { return (!path.empty() || mem.size); }
+        // returns true if path is not empty OR mem has a value set
+        explicit operator bool() const noexcept { return (!path.empty() || mem.size); }
 
-        operator const char*() { return path.c_str(); }
-        operator const gnutls_datum_t*() { return &mem; }
-        operator gnutls_x509_crt_fmt_t() { return ext; }
+        // Hidden behind a template so that implicit conversion to pointer doesn't cause trouble
+        template <typename T, typename = std::enable_if_t<std::is_same_v<T, gnutls_datum_t>>>
```
Comment (dr7ana) [inline include/quic/gnutls_crypto.hpp:71]: This was a good idea. Was there a specific issue prompting this change, or is this prophylactic?

File: include/quic/utils.hpp [hunk 2]
```diff
@@ -57,6 +61,13 @@ namespace oxen::quic
     using bstring_view = std::basic_string_view<std::byte>;
     namespace log = oxen::log;
 
+    constexpr bool IN_HELL =
```
Comment (dr7ana) [inline include/quic/utils.hpp:64]: 10/10

File: include/quic/utils.hpp [hunk 3]
```diff
@@ -455,27 +466,52 @@ namespace oxen::quic
         explicit io_result(int errno_val) : error_code{errno_val} {}
 
         // Constructs an io_result with an ngtcp2 error value.
-        io_result(int err, ngtcp2_error_code_t) : error_code{err}, is_ngtcp2{true} {}
+        static io_result ngtcp2(int err)
```
Comment (dr7ana) [inline include/quic/utils.hpp:469]: Minor: maybe renaming these to something like `ngtcp2_err` and `wsa_err` would be a bit clearer?

File: include/quic/utils.hpp [hunk 4]
```diff
@@ -583,7 +619,7 @@ namespace std
             }
 
             auto h = hash<string_view>{}(addr_data);
-            h ^= hash<in_port_t>{}(port) + inverse_golden_ratio + (h << 6) + (h >> 2);
+            h ^= hash<decltype(port)>{}(port) + inverse_golden_ratio + (h << 6) + (h >> 2);
```
Comment (dr7ana) [inline include/quic/utils.hpp:622]: Educational question: Previously the hash type was explicit to match the declaration of `port`. Why the change to `decltype`? Is this for platform compatibility?

**Review by dr7ana — COMMENTED** [review]
looks great, left a few questions about things I thought i was curious about

File: include/quic/gnutls_crypto.hpp (hunk 1 above)
Comment (jagerman) [reply include/quic/gnutls_crypto.hpp:71]: Not currently causing a problem AFAIK, but implicit conversion to a pointer type is all sorts of shotgun-y (because it allows things like `if (foo)` unintentionally).

File: include/quic/utils.hpp (hunk 3 above)
Comment (jagerman) [reply include/quic/utils.hpp:469]: They are static functions, so they always have to be prefixed with the `io_result::ngtcp(123)`

File: include/quic/utils.hpp (hunk 4 above)
Comment (jagerman) [reply include/quic/utils.hpp:622]: Because f-ing Windows has half-baked posix support for things like `in_port_t` that are supposed to be provided by aren't.
